src/main/java/com/mraof/minestuck/item/weapon/MagicAttackRightClickEffect.java

```diff
@@ -0,0 +1,186 @@
+	public static final MagicAttackRightClickEffect STANDARD_MAGIC = new MagicAttackRightClickEffect(15, 3, null, null, 1.0F, () -> ParticleTypes.ENCHANT, false);
+	public static final MagicAttackRightClickEffect POOL_CUE_MAGIC = new MagicAttackRightClickEffect(18, 4, null, null, 1.0F, () -> new RedstoneParticleData(20F, 0F, 0F, 2F), false);
+	public static final MagicAttackRightClickEffect HORRORTERROR_MAGIC = new MagicAttackRightClickEffect(20, 5, () -> new EffectInstance(Effects.WITHER, 100, 2), () -> MSSoundEvents.ITEM_GRIMOIRE_USE, 1.2F, () -> ParticleTypes.SQUID_INK, true);
+	public static final MagicAttackRightClickEffect ZILLY_MAGIC = new MagicAttackRightClickEffect(30, 8, null, null, 1.0F, () -> new RedstoneParticleData(20F, 20F, 20F, 2F), true);
+	public static final MagicAttackRightClickEffect ECHIDNA_MAGIC = new MagicAttackRightClickEffect(50, 8, null, null, 1.0F, () -> ParticleTypes.END_ROD, true);
+	
+	public MagicAttackRightClickEffect(int distance, int damage, Supplier<EffectInstance> effect, Supplier<SoundEvent> sound, float pitch, Supplier<IParticleData> particle, boolean explosiveFinish)
+	{
+		this.distance = distance;
+		this.damage = damage;
+		this.effect = effect;
+		this.sound = sound;
+		this.pitch = pitch;
+		this.particle = particle;
+		this.explosiveFinish = explosiveFinish;
+	}
+	
+	@Override
+	public ActionResult<ItemStack> onRightClick(World world, PlayerEntity player, Hand hand)
+	{
+		ItemStack itemStackIn = player.getHeldItem(hand);
+		
+		magicAttack(world, player, itemStackIn);
+		
+		if(player.isCreative())
+			player.getCooldownTracker().setCooldown(itemStackIn.getItem(), 10);
+		else
+			player.getCooldownTracker().setCooldown(itemStackIn.getItem(), 50);
+		
+		player.swing(hand, true);
+		itemStackIn.damageItem(6, player, playerEntity -> playerEntity.sendBreakAnimation(Hand.MAIN_HAND));
+		player.addStat(Stats.ITEM_USED.get(itemStackIn.getItem()));
+		return ActionResult.resultPass(itemStackIn);
+	}
+	
+	void magicAttack(World world, PlayerEntity player, ItemStack itemStack)
+	{
+		if(sound != null && player.getRNG().nextFloat() <= .1F) //optional sound effect adding
+			world.playSound(null, player.getPosX(), player.getPosY(), player.getPosZ(), sound.get(), SoundCategory.PLAYERS, 0.7F, pitch);
+		world.playSound(null, player.getPosX(), player.getPosY(), player.getPosZ(), SoundEvents.ENTITY_EVOKER_CAST_SPELL, SoundCategory.PLAYERS, 1.0F, 1.6F);
+		Vec3d vec3dPath;
+		Vec3d vecPos;
+		BlockPos blockPos;
+		
+		if(itemStack.getItem() == MSItems.ARTIFUCKER)
+		{
+			Vec3d randomFacingVecPos = new Vec3d(player.getPosX() + player.getRNG().nextInt(10) - 5, player.getPosY() + player.getRNG().nextInt(10) - 5, player.getPosZ() + player.getRNG().nextInt(10) - 5);
+			player.lookAt(player.getCommandSource().getEntityAnchorType(), randomFacingVecPos);
+		}
+		if(itemStack.getItem() == MSItems.POINTER_WAND)
```

Comment:
Instead of these item checks, I'd suggest providing these effects in some different way, such as using polymorphism with the effect class.

Reply:
Would another parameter, an integer that is 0 for the noneffected magic types, be acceptable as a way to test for these conditions?

Reply:
For example: 
new MagicAttackRightClickEffect(15, 8, null, null, 1.0F, () -> ParticleTypes.ENCHANTED_HIT, false, 1)

(last value is the added one, the value of 1 means it will have aimbot functionality)

Reply:
Sort of, but it should be an enum to describe the type instead of an int in that case.
However instead of a parameter, how does it sound if I make a pull request to you with a polymorphism solution?